dockstore-webservice/src/main/java/io/dockstore/webservice/core/metrics/MetricsDataS3Client.java

```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    /**
+     * This generates the s3 key of the metrics data that will be stored on S3.
+     *
+     * @param toolId       The GA4GH Tool ID
+     * @param versionName  The GA4GH ToolVersion name
+     * @param platform      The platform that the metrics data is from
+     * @param fileName     The file name to use. Should be the time that the data was submitted in milliseconds since epoch appended with '.json'
+     * @return S3 key (file path)
+     * @throws UnsupportedEncodingException
+     */
+    static String generateKey(String toolId, String versionName, String platform, String fileName) throws UnsupportedEncodingException {
+        List<String> pathList = new ArrayList<>();
+        pathList.add(S3ClientHelper.convertToolIdToPartialKey(toolId));
+        pathList.add(URLEncoder.encode(versionName, StandardCharsets.UTF_8));
+        pathList.add(URLEncoder.encode(platform, StandardCharsets.UTF_8));
+        pathList.add(URLEncoder.encode(fileName, StandardCharsets.UTF_8));
+        return String.join("/", pathList);
+    }
+
+    /**
+     * Get the metrics data JSON string from S3 for a GA4GH tool version
+     * @param toolId The GA4GH Tool ID
+     * @param versionName The GA4GH ToolVersion name
+     * @param platform The platform that the metrics data is from
+     * @param filename The file name of the S3 object
+     * @return JSON string containing the metrics data
+     * @throws IOException
+     */
+    public String getMetricsDataFileContent(String toolId, String versionName, String platform, String filename)
+            throws IOException {
+        String key = generateKey(toolId, versionName, platform, filename);
+        GetObjectRequest request = GetObjectRequest.builder().bucket(bucketName).key(key).build();
+        ResponseInputStream<GetObjectResponse> object = s3.getObject(request);
+        return IOUtils.toString(object, StandardCharsets.UTF_8);
+
+    }
+
+    /**
+     * Get a list of MetricsData for a GA4GH tool version
+     * @param toolId The GA4GH Tool ID
+     * @param toolVersionName The GA4GH ToolVersion name
+     * @return A list of MetricsData
+     * @throws UnsupportedEncodingException
+     */
+    public List<MetricsData> getMetricsData(String toolId, String toolVersionName) throws UnsupportedEncodingException {
+        String key = S3ClientHelper.convertToolIdToPartialKey(toolId) + "/" + URLEncoder.encode(toolVersionName, StandardCharsets.UTF_8);
+        ListObjectsV2Request request = ListObjectsV2Request.builder().bucket(bucketName).prefix(key).build();
+        ListObjectsV2Response listObjectsV2Response = s3.listObjectsV2(request);
+        List<S3Object> contents = listObjectsV2Response.contents();
+        return contents.stream().map(s3Object -> {
+            HeadObjectRequest build = HeadObjectRequest.builder().bucket(bucketName).key(s3Object.key()).build();
```

Comment:
This risks being slow if we do ever have the thousands of objects -- one API request per object.

But as I understand it, the idea is that this code will be called rarely -- the idea is we would only call it if we want to re-aggregate, right? And we won't have thousands of objects for a while, if ever.

Some thoughts:

* I'm overthinking this; don't optimize prematurely
* I think you can derive all the other fields in MetricsData from the key, without the metadata. You could do that, and then surface a new method to get the owner.
* Could you put the owner in the key? If you can do a wildcard listing of objects (?), then maybe you don't need the metadata at all.
* Generally we're going to treat the data as anonymous, so I suspect most cases, we won't need/want the owner.

Reply:
Yeah, this would only be called if we want to re-aggregate. 

> I think you can derive all the other fields in MetricsData from the key, without the metadata. You could do that, and then surface a new method to get the owner.

Yeah, that would work and would be more efficient. Like you mentioned, I suspect we will rarely need to know the owner, maybe in cases where someone is submitting malicious data and/or we want to delete all data in S3 from a certain user. I'll try to implement this suggestion